errors/errors.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package errors provides an error type for Go Cloud APIs.
+package errors
+
+import "fmt"
+
+// A Code describes the error's category. Programs should act upon an error's
+// code, not its message.
+type Code int
+
+const (
+	// The error could not be categorized.
+	Unknown Code = iota
+
+	// The resource was not found.
+	NotFound
+
+	// The resource exists, but it should not.
+	AlreadyExists
+
+	// The caller is not authorized to perform the operation.
+	PermissionDenied
+
+	// A value given to a Go Cloud API is incorrect.
```

Comment:
Are we committing to actually converting underlying provider errors to each of these? So far I believe we just have `NotFound`, adding all of these others might not be needed (yet).

Reply:
In my docstore prototype I needed AlreadyExists. InvalidArgument is important so callers can distinguish their own errors. E.g.  "blob.NewWriter: WriterOptions.Metadata keys may not be empty strings" should be InvalidArgument. Internal (better name?) is important because it indicates where _we_ messed up—users know to file a bug if they see it. The rest we don't need for now.

Reply:
OK, that sounds fine. We should document that driver implementations should return `Unknown` as a default.